crates/telio-firewall/src/firewall.rs

```diff
@@ -938,6 +986,202 @@ impl StatefullFirewall {
+        ipv4pkg.set_ttl(0xFF);
+        ipv4pkg.set_next_level_protocol(IpNextHeaderProtocols::Tcp);
+
+        let mut ipv4pkgbuf = [0u8; IPV4_LEN + TCP_LEN];
+
+        let iter = tcp_conn_cache.iter().filter_map(|(k, v)| {
+            if k.pubkey == *pubkey {
+                Some((&k.link, v))
+            } else {
+                None
+            }
+        });
+
+        for (key, val) in iter {
+            let Some(seq) = val.next_seq else {
+                // When we don't have the sequence number that means the
+                // connection is half open and we don't have a response
+                // from the remote peer yet. We need to skip this kind of
+                // connections
+                continue;
+            };
+
+            tcppkg.set_source(key.remote_port);
+            tcppkg.set_destination(key.local_port);
+            tcppkg.set_sequence(seq);
+
+            match (key.local_addr, key.remote_addr) {
+                (IpAddr::Ipv4(local_addr), IpAddr::Ipv4(remote_addr)) => {
+                    let src = remote_addr.into();
+                    let dst = local_addr.into();
+
+                    ipv4pkg.set_source(src);
+                    ipv4pkg.set_destination(dst);
+
+                    tcppkg.set_checksum(pnet_packet::tcp::ipv4_checksum(
+                        &tcppkg.to_immutable(),
+                        &src,
+                        &dst,
+                    ));
+
+                    ipv4pkg.set_checksum(0);
+                    ipv4pkg.set_checksum(pnet_packet::ipv4::checksum(&ipv4pkg.to_immutable()));
+
+                    telio_log_debug!("Injecting IPv4 TCP RST packet {key:#?}");
+
+                    #[allow(index_access_check)]
+                    ipv4pkgbuf[..IPV4_LEN].copy_from_slice(ipv4pkg.packet());
+
+                    #[allow(index_access_check)]
+                    ipv4pkgbuf[IPV4_LEN..].copy_from_slice(tcppkg.packet());
+
+                    sink4.write_all(&ipv4pkgbuf)?;
+                }
+                (IpAddr::Ipv6(_), IpAddr::Ipv6(_)) => (), // TODO(msz): implement this pice when IPv4 will be fully supported
+                _ => telio_log_warn!(
+                    "Local and remote IP addrs version missmatch, this should never happen"
+                ),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn reset_udp_conns(
+        &self,
+        pubkey: &PublicKey,
+        endpoint_ip: Option<StdIpAddr>,
+        sink4: &mut dyn io::Write,
+    ) -> io::Result<()> {
+        let Ok(mut udp_conn_cache) = self.udp.lock() else {
+            telio_log_error!("UDP cache poisoned");
+            return Ok(());
+        };
+
+        telio_log_debug!(
+            "Inspecting {} UDP connections for reset",
+            udp_conn_cache.len()
+        );
+
+        const IPV4_HEADER_LEN: usize = 20;
+        const ICMP_HEADER_LEN: usize = 8;
+        const ICMP_DATA_LEN: usize = UdpConnectionInfo::LAST_PKG_HEADER;
```

Comment:
here we must take care to see whether (`last_out_pkg_headers` did not have any IP options set. As otherwise it may be longer than 20 bytes, meaning, that full ICMP payload may be larger then 28 bytes.

Reply:
Done